types/topology.go

```diff
@@ -0,0 +1,302 @@
+package types
+
+import (
+	"os"
+	"path/filepath"
+
+	"github.com/docker/go-connections/nat"
+	"github.com/mitchellh/go-homedir"
+	"github.com/srl-labs/containerlab/utils"
+)
+
+// Topology represents a lab topology
+type Topology struct {
+	Defaults *NodeDefinition            `yaml:"defaults,omitempty"`
+	Kinds    map[string]*NodeDefinition `yaml:"kinds,omitempty"`
+	Nodes    map[string]*NodeDefinition `yaml:"nodes,omitempty"`
+	Links    []*LinkConfig              `yaml:"links,omitempty"`
+}
+
+func NewTopology() *Topology {
+	return &Topology{
+		Defaults: new(NodeDefinition),
+		Kinds:    make(map[string]*NodeDefinition),
+		Nodes:    make(map[string]*NodeDefinition),
+		Links:    make([]*LinkConfig, 0),
+	}
+}
+
+type LinkConfig struct {
+	Endpoints []string
+	Labels    map[string]string `yaml:"labels,omitempty"`
+}
+
+func (t *Topology) GetDefaults() *NodeDefinition {
+	if t.Defaults != nil {
+		return t.Defaults
+	}
+	return new(NodeDefinition)
+}
+
+func (t *Topology) GetKind(kind string) *NodeDefinition {
+	if t.Kinds == nil {
+		return new(NodeDefinition)
+	}
+	if kdef, ok := t.Kinds[kind]; ok {
+		return kdef
+	}
+	return new(NodeDefinition)
+}
+
+func (t *Topology) GetKinds() map[string]*NodeDefinition {
+	if t.Kinds == nil {
+		return make(map[string]*NodeDefinition)
+	}
+	return t.Kinds
+}
+
+func (t *Topology) GetNodeKind(name string) string {
+	if t == nil {
+		return ""
+	}
+	if ndef, ok := t.Nodes[name]; ok {
+		if ndef.GetKind() != "" {
+			return ndef.GetKind()
+		}
+		return t.GetDefaults().GetKind()
+	}
+	return ""
+}
+
+func (t *Topology) GetNodeBinds(name string) []string {
+	if ndef, ok := t.Nodes[name]; ok {
+		if len(ndef.GetBinds()) > 0 {
+			return ndef.GetBinds()
+		}
+		if len(t.GetKind(t.GetNodeKind(name)).GetBinds()) > 0 {
+			return t.GetKind(t.GetNodeKind(name)).GetBinds()
+		}
+		return t.GetDefaults().GetBinds()
+	}
+	return nil
+}
+
+func (t *Topology) GetNodePorts(name string) (nat.PortSet, nat.PortMap, error) {
+	if ndef, ok := t.Nodes[name]; ok {
+		// node level ports
+		if len(ndef.Ports) != 0 {
+			ps, pb, err := nat.ParsePortSpecs(ndef.Ports)
+			if err != nil {
+				return nil, nil, err
+			}
+			return ps, pb, nil
+		}
+		// kind level ports
+		if kdef, ok := t.Kinds[ndef.Kind]; ok && kdef != nil {
+			if len(kdef.Ports) > 0 {
+				ps, pb, err := nat.ParsePortSpecs(kdef.Ports)
+				if err != nil {
+					return nil, nil, err
+				}
+				return ps, pb, nil
+			}
+		}
```

Comment:
can we use kind-level ports? since a container port is bound to a specific host port, it seems it won't be possible to define a kind-scoped port forwarding as this would fail to bind the same host port >1 times